### ISSUE TITLE
Error while placing a funnel

### ISSUE THREAD

**Issue by pierredavidbelanger** [CleverRaven/Cataclysm-DDA#21586]
I am using a custom compiled version ( based on beeff69 + [ChezzoSounds](https://github.com/Chezzo/Cataclysm-DDA/tree/ChezzoSounds)) .

I placed (`%`) a funnel on a plastic bottle and got this message:

![cataclysm__dark_days_ahead_-_0_c-23226-ga2380b5](https://user-images.githubusercontent.com/1076726/29122985-33d7d9aa-7ce2-11e7-9c70-018c459f7a1d.png)

I was able to press space and continue to play.
The funnel was removed correctly from my inventory and placed on the bottle.

I saved and re-opened the game, everything is ok.

**Comment by pierredavidbelanger** [comment]
I am still able to reproduce this with the latest master (at commit 7f8c2e7).

I is easy: just wish for a funnel and place it

**Comment by angoddu** [comment]
Can not confirm this. I just wished myself in a funnel and it placed just fine

**Comment by pierredavidbelanger** [comment]
ok, I just pulled the latest master branch (at 4c45a9adbf895948b95b12e0564942500ec7b63e righrt now).

then 
```
$ uname -a
Darwin MacBook-Pro-de-Pierre-David.local 16.7.0 Darwin Kernel Version 16.7.0: Thu Jun 15 17:36:27 PDT 2017; root:xnu-3789.70.16~2/RELEASE_X86_64 x86_64
$ make clean
$ make app -j4 NATIVE=osx OSX_MIN=10.12 TILES=1 SOUND=1 LOCALIZE=0 RELEASE=1 CLANG=clang++ MACPORTS=1 USE_HOME_DIR=1
$ open Cataclysm.app
```

I created a world with every default options.

In this world I created a custom character with every default options.

Opened the debug menu, then `1` to Wish for an item, then find a `/funnel`, and gave myself `1` copy of it.

The wish is granted, I `ESC` to return to the game.

Then, with the Execute action menu (`%`), I selected `Place funnel`, followed by right direction (`l`)

Boom. again:

![cataclysm__dark_days_ahead_-_0_c-23286-gb0cf1d1](https://user-images.githubusercontent.com/1076726/29188155-38a321ac-7de0-11e7-86b5-d85ce650d774.png)

I am not sure what other information I can give to help with this :)

**Comment by angoddu** [comment]
Aah ok. I'm using the launcher on windows so I don't do my own compiles

**Comment by Treah** [comment]
I can confirm this occurs. It does not happen if you use the action menu and choose the funnel but using the % it does. 

[a] -> pick funnell -> choose direction -> places without error 
[%] -> place option -> choose direction - Error!

**Comment by AMurkin** [comment]
Can confirm on Win x86 Build 6669 by [%] menu.

**Comment by ZehCastro** [comment]
The item action menu path tries to remove the item twice.

First at `bool player::consume_charges( item& used, long qty )` because, according to the comment:  `// Tools which don't require ammo are instead destroyed`

Then at `void game::item_action_menu()`, because, again, according to the commend: `// Need to remove item` if invoking the action succeeded. Here the pointer to the item is not null but is invalid.

I still don't know enough about the code to tell if this second removal should be done at all, but it seems we cannot trust that the passed item pointer will still be valid after invoking the action.

Will keep digging later, at least to find out how the regular use path works.

**Comment by ZehCastro** [comment]
I think `void game::item_action_menu()` should not try to remove the item anymore. Even the removal was needed, the same bug would probably happen given the same conditions.

Right now `bool player::invoke_item( item* used, const std::string &method )` only returns true when `bool player::consume_charges( item& used, long qty )` also returns true.

The paths that return true inside the `consume_charges` also remove the item with `item Character::i_rem(const item *it)`. That being the case, there would be no need to remove it again.

When the removal was [added](https://github.com/CleverRaven/Cataclysm-DDA/blob/223ba1aaeff70bc5ccb61968fa16530519d96327/src/item_action.cpp#L255), [`player::consume_charges`](https://github.com/CleverRaven/Cataclysm-DDA/blob/223ba1aaeff70bc5ccb61968fa16530519d96327/src/player.cpp#L10748) would also remove the item but not when [`player::consume_item`](https://github.com/CleverRaven/Cataclysm-DDA/blob/223ba1aaeff70bc5ccb61968fa16530519d96327/src/player.cpp#L9146) was called for food itens. That might be why the removal was added in the first place, but this is just speculation.

**Comment by angoddu** [comment]
This is also a bug with placing anything through the % menu. I just had it happen with me brazer

**Comment by ZehCastro** [comment]
Yes, there are more cases that would trigger the bug. Most that involve the item that is acted upon being consumed, either because it was actually consumed or because it was transformed/assembled into something else.

The fix on #21732 will solve them, but I'm still worried that it might break some code I did not check. If it does break something, it would manifest as an item not being properly consumed after usage, and that might be a worst bug than the current NPE, in terms of its effect on gameplay.

That item removal was added with some changes related to food consumption. Checking that kind of interaction is probably the next step I'll take, whenever I get the time to do so.

**Comment by ZehCastro** [comment]
It would be more accurate to say that the bug will happen whenever the item was a tool, a medication or a container and `player::consume_charges( item& used, long qty )` returns true, which it does when it removes the item, ensuring that it will be removed again when the execution goes back to `void game::item_action_menu()` scope.

Or something like that...
